component/common/src/main/java/org/wso2/carbon/extension/identity/emailotp/common/util/OneTimePasswordUtils.java

```diff
@@ -0,0 +1,258 @@
+ * module's 'org.wso2.carbon.identity.authenticator.emailotp.OneTimePassword' class.
+ * Any fixes here, should be reflected there as well.
+ */
+public class OneTimePasswordUtils {
+
+    // Each OTP digit(0-9) has a corresponding checksum digit as in the below array.
+    private static final int[] CHECKSUM_DIGITS = {0, 2, 4, 6, 8, 1, 3, 5, 7, 9};
+
+    private static final Log log = LogFactory.getLog(OneTimePasswordUtils.class);
+
+    /**
+     * This method to generate a random number by giving the range.
+     *
+     * @param size length of the random number.
+     * @return random number string.
+     */
+    public static String getRandomNumber(int size) {
+
+        StringBuilder generatedToken = new StringBuilder();
+        try {
+            SecureRandom number = SecureRandom.getInstance(Constants.ALGORITHM_NAME);
+            // Generate 20 integers 0..20
+            for (int i = 0; i < size; i++) {
+                generatedToken.append(number.nextInt(9));
+            }
+        } catch (NoSuchAlgorithmException e) {
+            log.error("Unable to find the Algorithm.", e);
+        }
+
+        return generatedToken.toString();
+    }
+
+    /**
+     * This is to calculate the check sum of a number.
+     *
+     * @param num    The number to calculate the checksum for.
+     * @param digits Number of significant places in the number.
+     * @return The checksum of num.
+     */
+    public static int calcChecksum(long num, int digits) {
+
+        boolean doubleDigit = true;
+        int total = 0;
+        while (0 < digits--) {
+            int digit = (int) (num % 10);
+            num /= 10;
+            if (doubleDigit) {
+                digit = CHECKSUM_DIGITS[digit];
+            }
+            total += digit;
+            doubleDigit = !doubleDigit;
+        }
+        int result = total % 10;
+        if (result > 0) {
+            result = 10 - result;
+        }
+        return result;
+    }
+
+    /**
+     * This method uses the JCE to provide the HMAC-SHA-1
+     * algorithm. HMAC computes a Hashed Message Authentication Code and in this
+     * case SHA1 is the hash algorithm used.
+     *
+     * @param keyBytes Bytes to use for the HMAC-SHA-1 key.
+     * @param text     Message or text to be authenticated.
+     * @return Generated HMAC-SHA byte array.
+     * @throws NoSuchAlgorithmException If no provider makes either HmacSHA1 or HMAC-SHA-1 digest
+     *                                  algorithms available.
+     * @throws InvalidKeyException      The secret provided was not a valid HMAC-SHA-1 key.
+     */
+    public static byte[] hmacShaGenerate(byte[] keyBytes, byte[] text) throws NoSuchAlgorithmException,
+            InvalidKeyException {
+
+        Mac hmacSha;
+        try {
+            hmacSha = Mac.getInstance(Constants.ALGORITHM_HMAC);
+        } catch (NoSuchAlgorithmException nsa) {
+            hmacSha = Mac.getInstance(Constants.ALGORITHM_HMAC_SHA);
+        }
```

Comment:
HmacSHA1() function, uses a hash algorithm that is considered weak. ref: cwe 327.
But don't think the use of a weaker algorithm would pose a threat here since the OTP is valid for a brief period of time. WDYT?
cc: @ruwanta

Reply:
This `OneTimePasswordUtils` is exact code use in `org.wso2.carbon.identity.authenticator.emailotp.OneTimePassword` class and here didn't changed OTP generation algorithm is to keep the things same as email-OTP authenticator. Also can not use the same class a dependency and had been discussed about this one previously.